### PR TITLE
Graph similarity

### PR DESCRIPTION
**Pull request by dorinapetra**

### CODE REVIEW

File: src/hu_nmt/data_augmentator/augmentators/graph_based_augmentator.py [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+        pbar = tqdm(total=sample_count)
+        size = 0
+        while len(sampled_index_pairs) < sample_count:
+            (x, y) = np.random.choice(len(items), 2, replace=False)
+            if GraphBasedAugmentator._is_similar(items[x].hun, items[y].hun, dep) and \
+                    GraphBasedAugmentator._is_similar(items[x].eng, items[y].eng, dep):
+                sampled_index_pairs.add((x, y))
+                change = len(sampled_index_pairs) - size
+                pbar.update(change)
+                size = len(sampled_index_pairs)
```
Comment (attilanagy234) [inline src/hu_nmt/data_augmentator/augmentators/graph_based_augmentator.py:45]: Are we sure this won't run for a very long time with inappropriate thresholding? Have we tested this on a larger dataset?

Comment (attilanagy234) [reply src/hu_nmt/data_augmentator/augmentators/graph_based_augmentator.py:45]: Can we put in some safeguards to avoid infinite loops? (either break after X unsuccesful samples or gradually increase the threshold if it's not able to sample for a given threshold?)

File: src/hu_nmt/data_augmentator/augmentators/graph_based_augmentator.py [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+    def _is_similar(src_graph: DependencyGraphWrapper, tgt_graph: DependencyGraphWrapper, dep):
+        src_subgraph = GraphBasedAugmentator._get_subgraph(src_graph, dep)
+        tgt_subgraph = GraphBasedAugmentator._get_subgraph(tgt_graph, dep)
+        return GraphBasedAugmentator.similarity.get_similarity_from_graphs(src_subgraph.graph, tgt_subgraph.graph) >= \
+            GraphBasedAugmentator.threshold
+
+    @staticmethod
+    def _get_subgraph(wrapper: DependencyGraphWrapper, dep: str) -> nx.DiGraph:
+        edges_with_type = wrapper.get_edges_with_property('dep', dep)
+        edges_with_type = edges_with_type[0]
```
Comment (attilanagy234) [inline src/hu_nmt/data_augmentator/augmentators/graph_based_augmentator.py:59]: This assumes that we always constrain that there will be one `obj` and one `nsubj`, but not more. Maybe worth a comment

File: src/hu_nmt/data_augmentator/dependency_parsers/stanza_dependency_parser.py [hunk 3]
```diff
@@ -41,7 +41,8 @@ def sentence_to_node_relationship_list(nlp_pipeline, sent: str) -> List[NodeRela
             token = word.to_dict()
             target_key = f'{token["text"].lower()}_{token["id"]}'
             target_postag = token['upos']
-            target_lemma = token['lemma']
+            #target_lemma = token['lemma']
+            target_lemma = token.get('lemma')
```
Comment (attilanagy234) [inline src/hu_nmt/data_augmentator/dependency_parsers/stanza_dependency_parser.py:45]: why did we need to change this to `.get()` and not for the other attributes? Do we not always have the lemma there?

File: src/hu_nmt/data_augmentator/entrypoints/run_subj_obj_augmentation.py [hunk 4]
```diff
@@ -55,6 +59,17 @@ def main(src_language, tgt_language, src_data_folder, tgt_data_folder, augmentat
                                            filter_same_pos_tag=filter_same_pos_tag,
                                            filter_for_noun_tags=filter_for_noun_tags)
 
+
+    if augmentation_type == 'ged' or augmentation_type == 'edge_mapper':
```
Comment (attilanagy234) [inline src/hu_nmt/data_augmentator/entrypoints/run_subj_obj_augmentation.py:63]: can we give the old method a name as well and check for that? If it's none of the 3 valid options, let's throw an exception

File: src/hu_nmt/data_augmentator/graph_mappers/edge_mapper.py [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+    def __init__(self):
+        self.dep_weights = defaultdict(lambda: 1)
+
+        self.dep_weights['nsubj'] = 3
+        self.dep_weights['obj'] = 3
+        self.dep_weights['nmod'] = 2
+        self.dep_weights['obl'] = 2
+        self.dep_weights['case'] = 2
+        self.dep_weights['amod'] = 2
+        self.dep_weights['advmod'] = 2
```
Comment (attilanagy234) [inline src/hu_nmt/data_augmentator/graph_mappers/edge_mapper.py:19]: can we document how these were calculated and add comments here? Maybe a short explanation and a link pointing to a notebook with the analysis is fine.

File: src/hu_nmt/data_augmentator/graph_mappers/edge_mapper.py [hunk 6]
```diff
@@ -0,0 +1,161 @@
+import networkx as nx
+from collections import defaultdict, Counter
+import nltk
+from typing import List
+
+from hu_nmt.data_augmentator.graph_mappers.graph_similarity_base import GraphSimilarityBase, Edge, Node
+
+
+class EdgeMapper(GraphSimilarityBase):
```
Comment (attilanagy234) [inline src/hu_nmt/data_augmentator/graph_mappers/edge_mapper.py:9]: Can you please add docstring to this class summarizing what this class uses for similarity?

File: src/hu_nmt/data_augmentator/graph_mappers/edge_mapper.py [hunk 7]
```diff
@@ -0,0 +1,161 @@
...
+                else:
+                    min_routes = self._get_max_cand_by_route((s1, d1, data1), max_cands, g1, g2)
+                    if len(min_routes) == 1:
+                        mapping[(s1, d1)] = (min_routes[0][0], min_routes[0][1])
+                        g2_edges.remove(min_routes[0])
+                    else:
+                        max_children = self._get_max_children((s1, d1, data1), max_cands, g1, g2)
+                        mapping[(s1, d1)] = (max_children[0][0], max_children[0][1])
+                        g2_edges.remove(max_children[0])
+        return mapping
```
Comment (attilanagy234) [inline src/hu_nmt/data_augmentator/graph_mappers/edge_mapper.py:48]: Can you please add comments to this function (and the more important ones below) explaining the steps? It's kinda hard to follow without knowing the maths.

File: src/hu_nmt/data_augmentator/graph_mappers/ged.py [hunk 8]
```diff
@@ -0,0 +1,92 @@
+import networkx as nx
+from networkx import graph_edit_distance
+from typing import Dict
+
+from hu_nmt.data_augmentator.dependency_parsers.dependency_parser_factory import DependencyParserFactory
+from hu_nmt.data_augmentator.graph_mappers.graph_similarity_base import GraphSimilarityBase
+
+
+class GED(GraphSimilarityBase):
```
Comment (attilanagy234) [inline src/hu_nmt/data_augmentator/graph_mappers/ged.py:9]: Can you add a docstring explaining the main logic and motivation between this class as well please?

File: src/hu_nmt/data_augmentator/graph_mappers/ged.py [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+    def get_similarity_from_graphs(self, graph1: nx.DiGraph, graph2: nx.DiGraph):
+        init_distance = 0
+
+        # if the root pos tags are not the same
+        if graph1.nodes[self._get_root(graph1)]['postag'] != graph2.nodes[self._get_root(graph2)]['postag']:
+            graph1.nodes[self._get_root(graph1)]['postag'] = graph2.nodes[self._get_root(graph2)]['postag']
+            init_distance += 1
+
+        dist = self.get_ged(graph1, graph2) + init_distance
+        max_dist = len(graph1.nodes) * 2 - 1 + 2 * len(graph2.nodes) - 1
```
Comment (attilanagy234) [inline src/hu_nmt/data_augmentator/graph_mappers/ged.py:83]: what's this formula?

File: src/hu_nmt/data_augmentator/graph_mappers/graph_similarity_base.py [hunk 10]
```diff
@@ -0,0 +1,17 @@
...
+
+class GraphSimilarityBase(ABC):
+
+    @abstractmethod
+    def get_similarity_from_graphs(self, src_graph: nx.DiGraph, tgt_graph: nx.DiGraph) -> float:
+        raise NotImplementedError
+
+    def _get_root(self, graph: nx.DiGraph):
+        node = [n for n, d in graph.in_degree() if d == 0][0]
+        return node
```
Comment (attilanagy234) [inline src/hu_nmt/data_augmentator/graph_mappers/graph_similarity_base.py:17]: Why is this function in this class? I think it should belong in the graph wrapper. In fact, it's already there: https://github.com/attilanagy234/syntax-augmentation-nmt/blob/main/src/hu_nmt/data_augmentator/wrapper/dependency_graph_wrapper.py#L17

**Review by attilanagy234 — CHANGES_REQUESTED** [review]
Great job with this, I added all my comments below (mostly minor things). I have one big concern, which is that none of the newly added logic is unit tested and I wouldn't like to merge this without proper test coverage. Especially because you added a lot of new logic (dependent on math formulas and such). The similarity calculations and the new swap sampling logic / swap needs to be tested to the very least.

File: src/hu_nmt/data_augmentator/dependency_parsers/stanza_dependency_parser.py (hunk 3 above)
Comment (dorinapetra) [reply src/hu_nmt/data_augmentator/dependency_parsers/stanza_dependency_parser.py:45]: yes, I got errors earlier as there was no lemma, should I change the rest to be sure?

File: src/hu_nmt/data_augmentator/graph_mappers/edge_mapper.py (hunk 5 above)
Comment (dorinapetra) [reply src/hu_nmt/data_augmentator/graph_mappers/edge_mapper.py:19]: The first two are not relevant anymore as we compare subgraphs of the nsubj and obj edge. Kornai suggested the rest as these are the important edge types.

File: src/hu_nmt/data_augmentator/dependency_parsers/stanza_dependency_parser.py (hunk 3 above)
Comment (attilanagy234) [reply src/hu_nmt/data_augmentator/dependency_parsers/stanza_dependency_parser.py:45]: I assumed we got all of these correctly, so I'm a bit surprised we don't always get the lemma. I think for the lemma it's fine, because we are not using it for anything rn if I'm correct. What we could do is to rewrite the lemma to use `get` and if there's no lemma, just return the token itself as the lemma (and still populate the lemma field). For other fields I think it can stay like this, we need to make sure we always get a POS tag and a deprel.

File: src/hu_nmt/data_augmentator/graph_mappers/edge_mapper.py (hunk 5 above)
Comment (attilanagy234) [reply src/hu_nmt/data_augmentator/graph_mappers/edge_mapper.py:27]: I see. Can we still put this as a comment? That we assume these are the important dependency relations in the subtrees of an nsubj and an obj? And we can remove the unused weights.

**Review by attilanagy234 — CHANGES_REQUESTED** [review]

File: src/hu_nmt/test/graph/edge_mapper_test.py [hunk 11]
```diff
@@ -0,0 +1,137 @@
...
+                             (dep1)    (dep1)
+                               /          \ 
+                              b            c
+                           [tag2]        [tag3]
+                           /             /    \ 
+                        (dep3)       (dep4)  (dep3)
+                         /             /        \ 
+                        d             e          f  
+                     [tag4]        [tag5]      [tag6]
+        """
```
Comment (attilanagy234) [inline src/hu_nmt/test/graph/edge_mapper_test.py:69]: really appreciate the effort put into this 🔥

File: src/hu_nmt/data_augmentator/dependency_parsers/stanza_dependency_parser.py [hunk 12]
```diff
@@ -41,7 +41,7 @@ def sentence_to_node_relationship_list(nlp_pipeline, sent: str) -> List[NodeRela
             token = word.to_dict()
             target_key = f'{token["text"].lower()}_{token["id"]}'
             target_postag = token['upos']
-            target_lemma = token['lemma']
+            target_lemma = token.get('lemma') if 'lemma' in token else target_key
```
Comment (attilanagy234) [inline src/hu_nmt/data_augmentator/dependency_parsers/stanza_dependency_parser.py:44]: This can be simpler:
```python
token.get('lemma', target_key)

```

File: src/hu_nmt/data_augmentator/dependency_parsers/stanza_dependency_parser.py [hunk 13]
```diff
@@ -51,9 +51,10 @@ def sentence_to_node_relationship_list(nlp_pipeline, sent: str) -> List[NodeRela
                 head = word_dicts[int(token['head']) - 1]
                 source_key = f'{head["text"].lower()}_{head["id"]}'
                 source_postag = head['upos']
-                source_lemma = head['lemma']
+                source_lemma = head.get('lemma') if 'lemma' in head else source_key
```
Comment (attilanagy234) [inline src/hu_nmt/data_augmentator/dependency_parsers/stanza_dependency_parser.py:54]: This can be simpler:
```python
head.get('lemma', source_key)

**Review by attilanagy234 — CHANGES_REQUESTED** [review]

**Review by attilanagy234 — APPROVED** [review]
LGTM 🔥
